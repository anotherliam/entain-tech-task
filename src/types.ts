export type RaceSummary = {
  race_id: string;
  race_name: string;
  advertised_start: {
    seconds: number;
  };
  meeting_name: string;
  race_number: number;
  category_id: string;
};
