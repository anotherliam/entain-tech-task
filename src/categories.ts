export const categories: Record<string, string> = {
  '9daef0d7-bf3c-4f50-921d-8e818c60fe61': 'Greyhound Racing',
  '161d9be2-e909-4326-8c2c-35ed71fb460b': 'Horse Racing',
  '4a2788f8-e825-4d36-9894-efd4baf1cfae': 'Harness Racing',
};

export const categoryIds = Object.keys(categories);

export const categoryIdToName = (id: string) => categories[id] || 'Unknown';
