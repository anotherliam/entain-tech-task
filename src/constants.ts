export const colors = {
  background: '#eee',
  card: '#fff',
  border: '#000',
  primary: '#8e24aa',
  secondary: '#2e7d32',
};

// Time to wait between fetches
export const fetchDebounce = 10;

// Number of seconds until a race should dissapear after it starts
export const timeUntilRemove = 60;

export const racingEndpoint = 'https://api.neds.com.au/rest/v1/racing/?method=nextraces&count=10';
