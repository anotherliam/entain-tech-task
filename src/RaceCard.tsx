import { StyleSheet, View, Text } from 'react-native';
import React from 'react';
import { RaceSummary } from './types';
import { categoryIdToName } from './categories';
import { colors } from './constants';

type Props = {
  race: RaceSummary;
  time: number;
};

const RaceCard = ({ race, time }: Props) => {
  const timeUntilStart = race.advertised_start.seconds - time;
  const seconds = (timeUntilStart % 60).toString().padStart(2, '0');
  const minutes = Math.floor(timeUntilStart / 60);
  const timeString = timeUntilStart > 0 ? `Starts in ${minutes}:${seconds}` : 'Race has started!';
  return (
    <View style={[styles.card, timeUntilStart >= 0 ? styles.cardUpcoming : styles.cardStarted]}>
      <Text style={styles.raceTitle}>
        {race.meeting_name}, Race Number {race.race_number}
      </Text>
      <Text>{timeString}</Text>
      <Text>{categoryIdToName(race.category_id)}</Text>
    </View>
  );
};

export default RaceCard;

const styles = StyleSheet.create({
  card: {
    // ios shadow style
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.2,
    shadowRadius: 1.41,
    // android shadow style
    elevation: 2,

    backgroundColor: colors.card,
    padding: 8,
    marginVertical: 8,
    borderRadius: 4,
    borderLeftWidth: 2,
  },

  cardStarted: {
    borderLeftColor: colors.secondary,
  },

  cardUpcoming: {
    borderLeftColor: colors.primary,
  },

  raceTitle: {
    fontSize: 16,
    fontWeight: 'bold',
  },
});
