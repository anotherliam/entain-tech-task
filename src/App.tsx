import React, { useCallback, useEffect, useState } from 'react';
import {
  SectionList,
  SafeAreaView,
  StatusBar,
  StyleSheet,
  Text,
  View,
  SectionListData,
  ActivityIndicator,
} from 'react-native';
import { Picker } from '@react-native-picker/picker';
import { colors, fetchDebounce, racingEndpoint, timeUntilRemove } from './constants';
import { categoryIds, categoryIdToName } from './categories';
import { RaceSummary } from './types';
import RaceCard from './components/RaceCard';
import unionBy from 'lodash/unionBy';

type APIResponse = {
  data: {
    next_to_go_ids: string[];
    race_summaries: Record<string, RaceSummary>;
  };
};

const getCurrentTimeInSeconds = () => Math.floor(Date.now() / 1000);

const App = () => {
  const [error, setError] = useState<string | null>(null);
  const [lastFetchTime, setLastFetchTime] = useState(0);
  const [category, setCategory] = useState<string>('');
  const [sortedRaceData, setSortedRaceData] = useState<RaceSummary[]>([]);
  const [time, setTime] = useState<number>(0);

  /**
   * Called both on mount and periodically to fetch data.
   */
  const updateRaces = useCallback(async () => {
    setLastFetchTime(getCurrentTimeInSeconds);
    try {
      const response = await fetch(racingEndpoint);
      const body: APIResponse = await response.json();
      // Merge old races and new ones together, removing any that are more than 1 minute ago
      setSortedRaceData((prevRaces) => {
        const newRaces = body.data.next_to_go_ids.map((id) => body.data.race_summaries[id]);
        return unionBy(newRaces, prevRaces, 'race_id').sort(
          // sort by time then by id (to prevent elements jumping around if they have the same start time)
          (a, b) =>
            a.advertised_start.seconds - b.advertised_start.seconds ||
            a.race_id.localeCompare(b.race_id),
        );
      });
    } catch (e) {
      console.warn(e);
      setError('Failed to fetch data');
    }
  }, []);

  /**
   * Update timer every 100ms
   */
  useEffect(() => {
    let timer = setInterval(() => setTime(getCurrentTimeInSeconds()), 100);
    return () => clearInterval(timer);
  }, [setTime]);

  /**
   * On mount, do initial fetch
   */
  useEffect(() => {
    updateRaces();
  }, [updateRaces]);

  // Create the data for the section list
  const firstRaceToShowIndex =
    sortedRaceData.findIndex((race) => time <= race.advertised_start.seconds + timeUntilRemove) ||
    0;
  const racesToShow = sortedRaceData
    .slice(firstRaceToShowIndex, firstRaceToShowIndex + 5)
    .filter((race) => category === '' || category === race.category_id);
  const sectionListData: SectionListData<RaceSummary>[] = [
    {
      title: 'Started Races',
      data: racesToShow.filter((race) => race.advertised_start.seconds <= time),
    },
    {
      title: 'Upcoming Races',
      data: racesToShow.filter((race) => race.advertised_start.seconds > time),
    },
  ];

  // Trigger a new fetch if the most recent item in sorted races has expired and if its been at least 10s since the last fetch
  if (time >= lastFetchTime + fetchDebounce && firstRaceToShowIndex !== 0) {
    updateRaces();
  }

  return (
    <SafeAreaView style={styles.appContainer}>
      <StatusBar barStyle={'light-content'} />

      <View style={styles.categoryList}>
        <Picker selectedValue={category} onValueChange={(newCategory) => setCategory(newCategory)}>
          <Picker.Item label="All Categories" value={''} />
          {categoryIds.map((categoryId) => (
            <Picker.Item key={categoryId} value={categoryId} label={categoryIdToName(categoryId)} />
          ))}
        </Picker>
      </View>

      {error !== null ? (
        <View style={styles.errorMessageContainer}>
          <Text>{error}</Text>
        </View>
      ) : sortedRaceData.length <= 0 ? (
        <ActivityIndicator size="large" color={colors.primary} />
      ) : (
        <SectionList
          sections={sectionListData}
          keyExtractor={(race) => race.race_id}
          renderItem={({ item }) => <RaceCard race={item} time={time} />}
          renderSectionHeader={({ section }) =>
            section.data.length >= 1 ? (
              <Text style={styles.sectionTitleText}>{section.title}</Text>
            ) : null
          }
        />
      )}
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  appContainer: {
    backgroundColor: colors.background,
    flex: 1,
    padding: 12,
  },

  sectionTitleText: {
    marginLeft: 8,
    fontSize: 14,
  },

  categoryList: {
    borderBottomWidth: StyleSheet.hairlineWidth,
    marginBottom: 12,
  },

  errorMessageContainer: {
    alignItems: 'center',
  },
});

export default App;
