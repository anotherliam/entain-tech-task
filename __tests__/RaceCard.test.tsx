import 'react-native';
import React from 'react';
import RaceCard from '../src/RaceCard';

import renderer from 'react-test-renderer';
import { RaceSummary } from '../src/types';

const testRaceData: RaceSummary = {
  race_id: 'test-race-id',
  race_name: 'test-race-name',
  race_number: 1,
  meeting_name: 'test-meeting-name',
  advertised_start: {
    seconds: 15000,
  },
  category_id: '161d9be2-e909-4326-8c2c-35ed71fb460b',
};

it('matches snapshot in the future', () => {
  const tree = renderer.create(<RaceCard race={testRaceData} time={11000} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('matches snapshot in the past', () => {
  const tree = renderer.create(<RaceCard race={testRaceData} time={19000} />).toJSON();
  expect(tree).toMatchSnapshot();
});
