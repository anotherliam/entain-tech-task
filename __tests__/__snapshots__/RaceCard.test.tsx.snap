// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot in the future 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderLeftWidth": 2,
        "borderRadius": 4,
        "elevation": 2,
        "marginVertical": 8,
        "padding": 8,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 1.41,
      },
      Object {
        "borderLeftColor": "#8e24aa",
      },
    ]
  }
>
  <Text
    style={
      Object {
        "fontSize": 16,
        "fontWeight": "bold",
      }
    }
  >
    test-meeting-name
    , Race Number 
    1
  </Text>
  <Text>
    Starts in 66:40
  </Text>
  <Text>
    Horse Racing
  </Text>
</View>
`;

exports[`matches snapshot in the past 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderLeftWidth": 2,
        "borderRadius": 4,
        "elevation": 2,
        "marginVertical": 8,
        "padding": 8,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 1.41,
      },
      Object {
        "borderLeftColor": "#2e7d32",
      },
    ]
  }
>
  <Text
    style={
      Object {
        "fontSize": 16,
        "fontWeight": "bold",
      }
    }
  >
    test-meeting-name
    , Race Number 
    1
  </Text>
  <Text>
    Race has started!
  </Text>
  <Text>
    Horse Racing
  </Text>
</View>
`;
